active/0000-engines.md

```diff
@@ -0,0 +1,411 @@
+instance to interrogate its parent, specifically through its `RegistryProxy` and
+`ContainerProxy` interfaces.
+
+Alternatively, declarative dependencies can be defined on a limited basis. The
+initial API will be limited: an engine can define an array of `services` that it
+requires from its parent.
+
+For example, the following engine expects its parent to provide `store` and
+`session` services:
+
+```
+import Ember from 'ember';
+
+var Engine = Ember.Engine.extend({
+  dependencies: {
+    services: [
+      'store',
+      'session'
+    ]
+  }
+});
+
+export default Engine;
+```
+
+The parent application can provide a re-mapping of services from its namespace
+to that of the engine via an `engines` declaration.
+
+In the following example, the application shares its `store` service directly
+with the `checkout` engine. It also shares its `current-user` service as the
+`session` service requested by the engine.
+
+```
+import Ember from 'ember';
+
+var App = Ember.Application.extend({
+  engines: {
+    checkout: {
+      dependencies: {
+        services: [
+          'store',
+          {session: 'current-user'}
+        ]
+      }
+    }
+  }
+});
+
+export default App;
+```
+
+When engines are instantiated, the listed dependencies will be looked up on
+the parent and made accessible within the engine.
+
+Note that the `engines` declaration provides further space to define
+characteristics about an engine, such as whether it should be eager or
+lazy-loaded, URLs for manifest files, etc.
+
+# Drawbacks
+
+This RFC introduces the new concept of engines, which increases the
+learning curve of the framework. However, I believe this issue is
+mitigated by the fact that engines are an opt-in packaging around
+existing concepts.
+
+In the end, I believe that "engines" are just a small API for composing
+existing concepts. And they can be introduced at the top of the
+conceptual ladder, once users are comfortable with the basics of Ember,
+and only for those working on large teams or distributing addons.
+
+# Alternatives
+
+Several incomplete alternatives are discussed in the Motivations section above.
+
+I know of no alternatives being discussed in the Ember community that meet the
+same needs as engines; namely, for development _and_ run-time isolation.
+
+# Unresolved questions
+
+## Non-CLI Users
+
+This RFC assumes Ember CLI. I would prefer to prove this out in Ember
+CLI before locking down the public APIs/hooks the router exposes for
+locating and mounting engines. Once this is done, however, we should
+expose and document those hooks so users who cannot use Ember CLI for
+whatever reason can still take advantage of composability.
+
+## Declarative dependencies
+
+The initial scope of declarative dependency sharing is limited in scope to
+services. Should other types of dependencies be declaratively shareable?
+Should addons be the recommended path to share all other dependencies?
+
+## Async mounting of route-less engines
```

Comment:
I think there are some take aways here from the React Router 1.0.0-beta. It may also have some influence on other portions of this RFC. http://rackt.github.io/react-router/tags/v1.0.0-beta3.html#Advanced Usage